HDMI CEC support for XSqueeze

Snyder1977 opened the pull request:
These scripts enable XSqueeze to adjust the volume and the power mode of
an external hdmi avr-receiver via Ipeng or any other controller app.

Snyder1977 commented:
my first pull request... I hope that it will work....

bossanova808 commented:
Hey - sorry of the slow response.  Bit busy!

Ok - you need to sort out your line endings or whitespace issues - at the moment the diff is unreadable as it is showing you've changed every line of the entire thing, so I can't really see your changes.

I use windows, but I check in as unix style line endings I think.  Check your git settings.  The diff when you commit should show just the lines you have changed.  

I am no git expert, but I think you should:
Fix your local repo, squash it back down to one commit, and force push.  That should automatically update this pull request with the changes and alert me.  I will then review it.

Also - you will definitely need labels for those settings at the least!

bossanova808 commented:
I didn't get your above message notification for some reason.  Looks like you managed to get on top of it?

(I have contributed to several repos without any issues with endings, so you just need to sort your end out really - I don't think I want per repo settings files or anything, and I don't think this is generally necessary - I'm pretty sure it is standard practise to commit unix style and work with whatever you want at your end).

Anyway, now you just need to rebase as you're behind the head.  The push again and we should be good to review.  The end result should be your 'files changed' on github should show only your changes.

Almost there! 

A few other things I do notice though - constants use CAPITALS by convention, the settings page should have one master setting to enable/disable all the cec stuff so most people can ignore it (disables the other settings if in default state of off), and none of the cec stuff should default to on/true even when you enable the setting initially.

Snyder1977 commented:
I still haven´t figured out how to rebase. But I´ll keep on trying it.
Sorry for all the useless commits. It is my first time here on Github
so I´m not very experienced. I was just playing around with the github
client on my mac and was wondering how it interacts with the web
repository. I´m really sorry for the confusion.

I changed settings.xml which means that I added a master setting to
enable the hdmi support in general. I also switched all default hdmi
settings to false so that no hdmi commands will be send except someone
has enabled the cec support.

Are you fine with the line endings now? I think there are only 5 files
changed in comparison to your master branch. Can you now see all the
changes which I have made?

Snyder1977 commented:
Well, I probably need some help when it comes to rebasing. I have understood that I am behind like this example:
---------------------------------------------------------------------------------------------------------
Assume the following history exists and the current branch is "topic":

             A---B---C topic
            /
    D---E---F---G master

---------------------------------------------------------

Then I am supposed to enter this code:
git rebase master
git rebase master topic

The result of the commands should be:

                         A'--B'--C' topic
                        /
    D---E---F---G master

The problem is that I am always getting an error message:
fatal: Needed a single revision
invalid upstream Snyder1977/repository.bossanova808

Do you now what I do have to enter on the command line instead of "master"/"topic" in the code example?

bossanova808 commented:
Ok if it were me I would squash all the commits down to one commit, probably first.

Then rebase to master (I think I do this via a fetch from upstream and merge - I doubt any changes will clash but fix if they do) - the commit and force push to your upstream, that should update this to a clean single commit based on master.  

At the moment your commits are still removing some of my changes, so I can't pull them in.

You'll get there.  One other way would be to close this, re-clone from master, re-do your changes and go from there.  Given this is minor work it might just be easier than working out all the above (which I have done a few times but also have to look up each time - I use tortoisegit on windows but sometimes I just can't figure that out and do it command line style...)

Snyder1977 commented:
Sorry for the laste response. I am sure I´ll get there. I´m jstill quite busy at the moment so I need some time to figure out how this github stuff works. As you know I am a newbie ..... I have installed sourcetree on my mac now. This tool supports merging and rebasing. So I have a nice graphical interface which seems to be easier to use compared to the command line.

If I don´t manage to squash my commits into one commit and to rebase to master, I´ll close this and reclone from master. We will see. I´ll keep on trying. As soon as I´ll have the time I´ll get back to you.

Snyder1977 commented:
Can you merge it now?

bossanova808 commented on staging/script.xsqueeze/resources/lib/classes/NowPlayingWindow.py:
```diff
@@ -295,7 +302,41 @@ def update(self):
+          if powerstate == 0 and constants.HDMIPOWER == "true":
+            if constants.TURNOFFTV == "true":
+               tv.standby()
+               pass
```
not needed?

Snyder1977 replied:
I would say it is a nice feature to turn off the TV. Yet I didn´t add a feature which enables XSqueeze to turn on the TV because setting the TV-Input seems to be a little bit more difficult.

bossanova808 commented on staging/script.xsqueeze/resources/lib/classes/NowPlayingWindow.py:
```diff
@@ -295,7 +302,41 @@ def update(self):
+
+        if constants.ADJUSTVOL == "true" and constants.CECSUPPORT == 'true':
+          vol1 = self.player.getVolume()
+          xbmc.sleep(1000)
```
Why two calls to get volume, and why a sleep?  You're in a loop here, that operates basically every half a second anyway.  It then grabs a thread lock and should process everything quickly before sleeping the half second.

Snyder1977 replied:
Well, I have expected this question. Maybe it is not the best solution but it works :-) Here I want to find out if the controller app has turned up or turned down the volume. After putting the system to sleep the second call for volume delivers another value than the first one.

bossanova808 commented on staging/script.xsqueeze/resources/lib/classes/NowPlayingWindow.py:
```diff
@@ -399,6 +440,19 @@ def onClick( self, control ):
 
+         #hdmi support
+          if (control == constants.BUTTONVOLUP) and constants.CECSUPPORT == 'true':
+            cec.volume_up()
```
4 calls?  Why?  And why not a loop?

Snyder1977 replied:
A loop would be better. I will change it in my next commit. I will also link this loop with the volume factor (see settings.xml). The number of calls will depend on the volume factor then.

bossanova808 commented on staging/script.xsqueeze/resources/settings.xml:
```diff
@@ -45,4 +45,12 @@
     </category>
+        <category label="HDMI CEC">
+       <setting id="cecsupport" type="bool" label="19929" default="false"/>
+       <setting id="turnonavr" type="bool" label="19924" default="false"/>
```
You need a proper settings page with explanatory labels please.

Snyder1977 replied:
ok, I will provide some more explanations.

bossanova808 commented:
On github it is showing 15 separate commits.  It's also showing 15 files changed.  So you're still not there.  It should show 1 commit and ~5 files changed.

Anyway:
First, you should address the issues I have just raised - thanks.  (This is the first time I've looked at code as it's finally roughly a clean diff).

Then, honestly at this point I would simply re-close this PR, re-clone fresh from current master (you're still below head) and to that apply JUST your changes, and commit once locally and make sure it's a single commit etc.  Then, and only then, push it upstream to here as a new pull request.

I'm going to close this one for now anyway so it's off my to-do till it's in better shape.  You're pretty close though!  (Trust me if you ever do a commit to kodi itself or a major addon, you'll need to know all this and a lot more - my first one seemed like endless torture!!)

Snyder1977 commented:
Hi, it´s good to know that your first commit was a endless torture. Makes me feel a little bit better :-) I´m really sorry for bothering you with so many questions and useless commits/PRs....Anyway probably every newbie has to go through this....
